### PR TITLE
Console

### PR DESCRIPTION
**Pull request by lucasvanhalst**
See https://gitlab.com/ourcityapp_dev/app-configurator/-/merge_requests/1

TODO:
SolutionServerSettings
set `appcfg_server_url`
set `gsuite_service_account`
set `gsuite_delegated_user`
set `gsuite_domain`
set `default_email_group_members`

ServerSettings
set `emailHashEncryptionKey`

### CODE REVIEW

File: src/app.yaml [hunk 1]
```diff
@@ -207,22 +207,30 @@ handlers:
     secure: always
 
   - url: /restapi.*
-    script: rogerthat_unauthenticated.app
+    script: main_google_authenticated.app
```
Comment (RubenMattan) [inline src/app.yaml:210]: is this correct?

File: src/app.yaml [hunk 2]
```diff
@@ -207,22 +207,30 @@ handlers:
...
+    login: required
     secure: always
 
   - url: /images.*
-    script: rogerthat_unauthenticated.app
+    script: main_google_authenticated.app
```
Comment (RubenMattan) [inline src/app.yaml:215]: is this correct?

File: src/app.yaml [hunk 3]
```diff
@@ -207,22 +207,30 @@ handlers:
...
+    login: required
     secure: always
 
   - url: /uploads.*
-    script: rogerthat_unauthenticated.app
+    script: main_google_authenticated.app
```
Comment (RubenMattan) [inline src/app.yaml:220]: is this correct?

File: src/shop/view.py [hunk 4]
```diff
@@ -2333,3 +2332,23 @@ def rest_import_customers(import_id, app_id, file_data):
...
+        if DEBUG:
+            url = 'http://localhost:4199' + route
+        else:
+            url = '/console' + route
+        path = os.path.join(os.path.dirname(__file__), 'html', 'console.html')
+        context = get_shop_context(url=url)
+        self.response.out.write(template.render(path, context))
+
+
+class ConsoleIndexHandler(webapp2.RequestHandler):
```
Comment (RubenMattan) [inline src/shop/view.py:2348]: no need to authenticate?

File: src/main_google_authenticated.py [hunk 5]
```diff
@@ -63,9 +69,23 @@
...
+    webapp2.Route('/uploads/apps/<app_id:[^/]+>/assets/<asset_type:[^/]+>', UploadAppAssetHandler),
+    webapp2.Route('/uploads/assets', UploadGlobalAppAssetHandler),
+    webapp2.Route('/uploads/assets/<asset_id:[^/]+>', UploadGlobalAppAssetHandler),
+    webapp2.Route('/uploads/apps/<app_id:[^/]+>/default-brandings/<branding_type:[^/]+>', UploadDefaultBrandingHandler),
+    webapp2.Route('/uploads/default-brandings', UploadGlobalBrandingHandler),
+    webapp2.Route('/uploads/default-brandings/<branding_id:[^/]+>', UploadGlobalBrandingHandler),
+    webapp2.Route('/restapi/proxy<route:.*>', ProxyHandlerConfigurator),
 ]
 handlers.extend(rest_functions(view, authorized_function=authorize_manager))
+handlers.extend(rest_functions(apps))
```
Comment (RubenMattan) [inline src/main_google_authenticated.py:85]: no auth needed?

File: src/main_google_authenticated.py [hunk 6]
```diff
@@ -63,9 +69,23 @@
...
+handlers.extend(rest_functions(embedded_apps))
```
Comment (RubenMattan) [inline src/main_google_authenticated.py:86]: no auth needed?

File: src/main_google_authenticated.py [hunk 7]
```diff
@@ -63,9 +69,23 @@
...
+handlers.extend(rest_functions(firebase))
```
Comment (RubenMattan) [inline src/main_google_authenticated.py:87]: no auth needed?

File: src/main_google_authenticated.py [hunk 8]
```diff
@@ -63,9 +69,23 @@
...
+handlers.extend(rest_functions(payment))
```
Comment (RubenMattan) [inline src/main_google_authenticated.py:88]: no auth needed?

File: src/main_google_authenticated.py [hunk 9]
```diff
@@ -63,9 +69,23 @@
...
-    (shopOauthDecorator.callback_path, shopOauthDecorator.callback_handler())  # /shop/oauth2callback
+    (shopOauthDecorator.callback_path, shopOauthDecorator.callback_handler()),  # /shop/oauth2callback
+    webapp2.Route('/internal/console<route:.*>', ConsoleHandler),
+    webapp2.Route('/console<route:.*>', ConsoleIndexHandler),
+    webapp2.Route('/images/apps/<app_id:[^/]+>/qr-templates/<description:[^/]+>', AppQRTemplateHandler),
+    webapp2.Route('/uploads/apps/<app_id:[^/]+>/assets/<asset_type:[^/]+>', UploadAppAssetHandler),
+    webapp2.Route('/uploads/assets', UploadGlobalAppAssetHandler),
+    webapp2.Route('/uploads/assets/<asset_id:[^/]+>', UploadGlobalAppAssetHandler),
+    webapp2.Route('/uploads/apps/<app_id:[^/]+>/default-brandings/<branding_type:[^/]+>', UploadDefaultBrandingHandler),
+    webapp2.Route('/uploads/default-brandings', UploadGlobalBrandingHandler),
```
Comment (RubenMattan) [inline src/main_google_authenticated.py:80]: no auth needed?

File: src/app.yaml (hunk 1 above)
Comment (lucasvanhalst) [reply src/app.yaml:210]: absolutely

File: src/rogerthat/bizz/app.py [hunk 10]
```diff
@@ -369,6 +378,26 @@ def trans(create_app_to):
...
+        appcfg_data = create_app_to.to_dict()
+        server_settings = get_server_settings()
+        appcfg_data['app_constants'] = {
+            'EMAIL_HASH_ENCRYPTION_KEY': server_settings.emailHashEncryptionKey,
+            'REGISTRATION_EMAIL_SIGNATURE': server_settings.registrationEmailSignature,
+            'REGISTRATION_MAIN_SIGNATURE': server_settings.registrationMainSignature,
+            'REGISTRATION_PIN_SIGNATURE': server_settings.registrationPinSignature,
+            'GOOGLE_MAPS_KEY': server_settings.googleMapsKey,
+        }
+        appcfg_data['cloud_constants'] = {
```
Comment (RubenMattan) [inline src/rogerthat/bizz/app.py:394]: WIP

File: src/rogerthat/bizz/gsuite.py [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+from solution_server_settings import get_solution_server_settings
+
+SCOPES = ['https://www.googleapis.com/auth/admin.directory.group',
+          'https://www.googleapis.com/auth/admin.directory.group.readonly',
+          'https://www.googleapis.com/auth/admin.directory.group.member']
+
+
+def _get_admin_directory_service():
+    server_settings = get_solution_server_settings()
+    account = json.loads(server_settings.gsuite_service_account)
```
Comment (RubenMattan) [inline src/rogerthat/bizz/gsuite.py:39]: add to description for todo

File: src/rogerthat/bizz/gsuite.py [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+    credentials = ServiceAccountCredentials._from_parsed_json_keyfile(account, SCOPES)\
+        .create_delegated(server_settings.gsuite_delegated_user)
```
Comment (RubenMattan) [inline src/rogerthat/bizz/gsuite.py:41]: add to description for todo

File: src/rogerthat/bizz/gsuite.py [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+    http = credentials.authorize(httplib2.Http())
+    return discovery.build('admin', 'directory_v1', http=http)
+
+
+def create_app_group(app_id):
+    email = '%s@%s' % (app_id, get_solution_server_settings().gsuite_domain)
```
Comment (RubenMattan) [inline src/rogerthat/bizz/gsuite.py:47]: add to description for todo

File: src/rogerthat/bizz/gsuite.py [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+            raise
+        else:
+            logging.info('No new group added, group for app_id %s already exists', app_id)
+
+
+def _add_default_members_to_group(group_id):
+    # Use batch requests if this is too slow https://developers.google.com/api-client-library/python/guide/batch
+    service = _get_admin_directory_service()
+    server_settings = get_solution_server_settings()
+    members = json.loads(server_settings.default_email_group_members)
```
Comment (RubenMattan) [inline src/rogerthat/bizz/gsuite.py:64]: add to description for todo

File: src/solution_server_settings/__init__.py [hunk 15]
```diff
@@ -34,7 +34,21 @@ class SolutionServerSettings(CachedModelMixIn, db.Model):
     bob_api_secret = add_meta(db.StringProperty(indexed=False),
                               doc="The secret used for incoming api request",
                               order=2)
-
+    appcfg_server_url = add_meta(db.StringProperty(indexed=False),
```
Comment (RubenMattan) [inline src/solution_server_settings/__init__.py:37]: belongs to serversettings

File: src/solutions/common/bizz/participation/proxy.py [hunk 16]
```diff
@@ -136,11 +136,13 @@ def update_settings(service_user, data):
     return json.loads(_participation_request(model.secret, url, 'PUT', data).content)
 
 
-def register_app(app_id, ios_dev_team):
-    # type: (unicode, unicode) -> None
+def register_app(app_id, ios_developer_account_id):
+    # type: (unicode, long) -> None
+    raise NotImplementedError('TODO: fetch developer account from appcfg')
```
Comment (RubenMattan) [inline src/solutions/common/bizz/participation/proxy.py:141]: todo found

**Pull request merged**